src/libmw/src/file/File.cpp

```diff
@@ -0,0 +1,205 @@
+    if (parent == m_path) {
+        ThrowFile_F("Can't find parent path for {}", *this);
+    }
+
+    const FilePath destination = parent.GetChild(filename);
+    if (destination.Exists()) {
+        destination.Remove();
+    }
+
+    std::error_code ec;
+    filesystem::rename(m_path.m_path, destination.m_path, ec);
+    if (ec) {
+        ThrowFile_F("Failed to rename {} to {}", *this, destination);
+    }
+
+    m_path = destination.m_path;
+}
+
+std::vector<uint8_t> File::ReadBytes() const
+{
+    std::error_code ec;
+    if (!filesystem::exists(m_path.m_path, ec) || ec) {
+        ThrowFile_F("{} not found", *this);
+    }
+
+    std::ifstream file(m_path.m_path, std::ios::in | std::ios::binary);
+    if (!file.is_open()) {
+        ThrowFile_F("Failed to open {} for reading", *this);
+    }
+
+    const size_t size = (size_t)filesystem::file_size(m_path.m_path, ec);
+
+    std::vector<uint8_t> bytes((size_t)size);
+    file.seekg(0, std::ios::beg);
+    file.read((char*)bytes.data(), size);
+    file.close();
+
+    return bytes;
+}
+
+std::vector<uint8_t> File::ReadBytes(const size_t startIndex, const size_t numBytes) const
+{
+    std::error_code ec;
+    if (!filesystem::exists(m_path.m_path, ec) || ec) {
+        ThrowFile_F("{} not found", *this);
+    }
+
+    std::ifstream file(m_path.m_path, std::ios::in | std::ios::binary);
+    if (!file.is_open()) {
+        ThrowFile_F("Failed to open {} for reading", *this);
+    }
+
+    const size_t size = (size_t)filesystem::file_size(m_path.m_path, ec);
+    if (size < (startIndex + numBytes)) {
+        ThrowFile_F("Failed to read {} bytes from {}.", numBytes, *this);
+    }
+
+    std::vector<uint8_t> bytes((size_t)numBytes);
+    file.seekg(startIndex, std::ios::beg);
+    file.read((char*)bytes.data(), numBytes);
+    file.close();
+
+    return bytes;
+}
+
+void File::Write(const std::vector<uint8_t>& bytes)
+{
+    if (!Exists()) {
+        Create();
+    }
+
+    std::ofstream file(m_path.m_path, std::ios::out | std::ios::binary | std::ios::app);
+    if (!file.is_open()) {
+        ThrowFile_F("Failed to write to file: {}", m_path);
+    }
+
+    file.write((const char*)bytes.data(), bytes.size());
+    file.close();
+}
+
+void File::Write(const size_t startIndex, const std::vector<uint8_t>& bytes, const bool truncate)
+{
+    if (!bytes.empty()) {
+        std::fstream file(m_path.m_path, std::ios::out | std::ios::binary | std::ios::app);
+        if (!file.is_open()) {
+            ThrowFile_F("Failed to write to file: {}", m_path);
+        }
+
+        file.seekp(startIndex, std::ios::beg);
+        file.write((const char*)bytes.data(), bytes.size());
+        file.close();
+    }
+
+    if (truncate) {
+        Truncate(startIndex + bytes.size());
+    }
+}
+
+void File::WriteBytes(const std::unordered_map<uint64_t, uint8_t>& bytes)
+{
+    std::fstream file(m_path.m_path, std::ios_base::binary | std::ios_base::out | std::ios_base::in);
```

Comment:
Some of these flags should be tidied up a bit. Its mostly std::ios in this file. Also why does it need std::ios::in ?